Add i18n-no-newlines rule

deBhal opened the pull request:
This PR adds a new rule to catch newlines in translatable strings. We've seen newlines turning up in translatable strings where they shouldn't be related to the increasing use of literal templates, but not using newlines is a pretty good rule of thumb.

The README has a more complete justification.

In the process I noticed that the other i18n rules weren't handling templates, so I've been enthusiastic about breaking out functionality I think will be useful for those.

Commit:
Add i18n-no-newlines rule

aduth commented on lib/util/get-string-from-node.js:
```diff
@@ -0,0 +1,26 @@
+	// TemplateLiteral quasis =>  node.value.raw
+	// Literal strings => node.value
+	return getStringLeafNodes( node )
+		.map( ( leaf ) => leaf.value.raw || leaf.value )
```
I imagine using arrow functions imposes a certain minimum Node version requirement on consuming applications?

ntwb replied:
It does, NodeJS 4.x/ES6, currently this repo supports NodeJS `>=0.10.0` per [package.json](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/package.json#L29)

aduth commented on lib/util/get-string-leaf-nodes.js:
```diff
@@ -0,0 +1,26 @@
+function getStringLeafNodes( node ) {
+	if ( node.type === 'BinaryExpression' && node.operator === '+' ) {
+		return getStringLeafNodes( node.left ).concat( getStringLeafNodes( node.right ) );
+	};
```
Unnecessary semicolon

ntwb replied:
Maybe add the ESLint rule `no-extra-semi` to enforce this?

aduth commented on lib/util/get-string-leaf-nodes.js:
```diff
@@ -0,0 +1,26 @@
+	};
+
+	if ( node.type === 'Literal' && 'string' === typeof node.value ) {
+		return [ node ]
```
Necessary semicolon 😆

ntwb replied:
Maybe add the ESLint rule `semi` to enforce this? 😆

aduth commented on tests/lib/util/get-string-from-node.js:
```diff
@@ -0,0 +1,38 @@
+	var programNode = parser.parse( code, config.env );
+	// Espree thinks it's parsing a whole program, so we just need to peel away
+	// the 'Program' packaging.
+	var stringNode = programNode.body[0].expression;
```
Looks like we should update this project to use `eslint-config-wpcalypso`. We've since changes the linting rules to warn on this usage (rather, prefer spaces in array access, even when using number or string).

aduth commented on tests/lib/rules/i18n-no-newlines.js:
```diff
@@ -0,0 +1,57 @@
+			} ]
+		},
+		{
+			code: "translate( 'My double-quoted string\\nwith a newline' );",
```
Minor, but you have the single-quoted / double-quoted mixed up in these test cases.

ntwb replied:
This is expected because of the `avoid-escape` option in the `quotes` rule:
• `"quotes": [ 2, "single", "avoid-escape"`
• https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/.eslintrc.json#L50
• http://eslint.org/docs/rules/quotes

aduth replied:
> This is expected because of the `avoid-escape` option in the `quotes` rule:

Yeah, I think it's fine that the object value is a double-quoted string, just that the "My double-quoted string" part of the test isn't entirely accurate, since the string being parsed is single-quoted.

deBhal replied:
Whoops!  I do like these sorts of placeholder strings to be accurate and identifiable - it makes it so much easier to understand what's going on if the test ever fails.

aduth commented on lib/util/get-string-from-node.js:
```diff
@@ -0,0 +1,26 @@
+ */
+var getStringLeafNodes = require( './get-string-leaf-nodes' );
+
+function getStringFromNode( node ) {
```
I don't know that this handles interpolation very well:

```
> getStringFromNode( parser.parse( '`Hello ${ target }`' ).body[ 0 ].expression )
'Hello [object Object]'
```

deBhal replied:
I had a look at reconstructing the string in a template, and it's a bit interesting, but then I realized that any expression will foul up the translation system, so for the purposes of this module, you can assume there will never be more than one quasi, and no expressions.

This definitely warrants a note in the docstring.

aduth commented on lib/rules/i18n-no-newlines.js:
```diff
@@ -0,0 +1,46 @@
+	};
+};
+
+rule.ERROR_MESSAGE = 'No newlines in translations unless you REALLY mean it';
```
I think it's kinda a given that rules can be overridden, so not sure about the "unless you REALLY mean it" part. Wonder if instead we can imply that it's probably an accident or that translations should usually not contain newlines?

> Unintended newline in translation

or

> Translations usually don't contain newlines

What do you think? Seems an odd rule in this sense that it's trying to catch accidents while still allowing intended usage.

deBhal commented:
I like your "usually" formulation, so I've gone with that.  The ambiguity here is a bit odd, but I think it's justified.

It's not at all obvious that newlines are going to cause the amount of hassle that they do (typically queries from translators of multiple locales), and while it's not actually _wrong_ for newlines to be in a string that we call `translate()` on, it is exceptional.  It's much more normal for a string like that to be pulled out into a separate file.

I've also added the `semi` and `no-extra-semi` rules, which probably doesn't belong in this PR, but it's done now.

aduth commented:
> I've also added the `semi` and `no-extra-semi` rules, which probably doesn't belong in this PR, but it's done now.

Just saw @ntwb's comment on this now. We should probably create a pull request to extend [`eslint-config-wpcalypso`](https://github.com/Automattic/eslint-config-wpcalypso).

At which point, `eslint-plugin-wpcalypso` will extend `eslint-config-wpcalypso` which extends `eslint-plugin-wpcalypso`. 😆

deBhal commented:
Infinite recursion aside ;), that makes more sense than sticking it in here.  I've reverted the `.eslintrc.json` change and squashed, so I think we're good to go?

aduth commented:
Made another pass at this. Looks good to me 👍 

We'll want to add a note to the changelog and bump the version. Did you want to wait for #15 and/or the `eslint-config-wpcalypso` change to bump the version? In any case, I'd suggest next version including this should be a minor version bump.

Commit:
Bump version to 1.3.4

deBhal commented:
I've got other priorities ahead of it, so #15 might take a little while to get to.  There's no harm in it, so I'll bump the version now so I can wrap this up, but I'm happy to wait to include the config change before we deploy the new version and update calypso.

aduth commented:
Since it's adding a new backwards-compatible feature, I think the minor bump 1.4.0 would be more appropriate.

deBhal commented:
Yeah, that makes sense.

deBhal commented:
Seeing how I jumped the gun on the merge here I'll commit straight to master, because two wrongs make a right, as they say.

https://github.com/Automattic/eslint-plugin-wpcalypso/commit/7785acb1378090a8e0ef0dac23e874b556d0eb3f

aduth commented:
> I'll commit straight to master

Fine with me 😄
